data/transactions/logic/eval.go

```diff
@@ -859,13 +858,15 @@ func eval(program []byte, cx *EvalContext) (pass bool, err error) {
 	}
 
 	for (err == nil) && (cx.pc < len(cx.program)) {
-		if cx.Debugger != nil {
-			if derr := cx.Debugger.Update(cx.refreshDebugState(err)); derr != nil {
-				return false, derr
-			}
+		if cx.Tracer != nil {
+			cx.Tracer.BeforeOpcode(cx)
 		}
 
 		err = cx.step()
+
+		if cx.Tracer != nil {
+			cx.Tracer.AfterOpcode(cx, err)
+		}
```

Comment:
No need to do it now, but I wonder if it would be worth hoisting the `cx.Tracer != nil` out of the loop.

We check it, and if it's there, we use a loop that calls BeforeOpcode and AfterOpcode, else we use a loop with just `err = cx.step()`. Seems like a dumb little optimization, but it's out hot loop.

Reply:
Good point. As a follow up, I'd be interested in benchmarking to see whether there is a difference.